packages/jsii/lib/compiler.ts

```diff
@@ -1,1573 +1,178 @@
+};
+
+const LOG = log4js.getLogger('jsii/compiler');
+export const DIAGNOSTICS = 'diagnostics';
+
+export interface CompilerOptions {
+    /** The information about the project to be built */
+    projectInfo: ProjectInfo;
+    /** Whether the compiler should watch for changes or juts compile once */
+    watch: boolean;
+}
 
-        function flagNames(t: ts.Type): string[] {
-            const result = new Array<string>();
-            for (const flag of Object.keys(ts.TypeFlags)) {
-                // Number-enums have item indexes in Object.keys
-                if (!isNaN(Number(flag))) { continue; }
-                const flagValue = (ts.TypeFlags as any)[flag];
-                // tslint:disable-next-line:no-bitwise
-                if (t.getFlags() & flagValue) {
-                    result.push(flag);
-                }
-            }
-            return result;
-        }
-    }
+export class Compiler implements Emitter {
+    public constructor(private readonly options: CompilerOptions) {}
 
     /**
-     * Given a type symbol, returns a qualified type name.
+     * Compiles the configured program.
      */
-    async function getFullyQualifiedName(typeSymbol: ts.Symbol, ctx: string[]) {
-        ctx = newContext(ctx, typeSymbol.name);
-
-        const fqn = typeChecker.getFullyQualifiedName(typeSymbol);
-        const groups = /\"([^\"]+)\"\.(.*)/.exec(fqn);
-        if (!groups) {
-            throw error(ctx, `Cannot use unexported type '${fqn}' in a public API`);
-        }
-
-        const modulePath = groups[1];
-        const typeName = groups[2];
-
-        const packageConfig = await findPackageConfig(modulePath);
-        if (!packageConfig) {
-            throw error(ctx, 'Cannot find package.json up the tree from: ' + modulePath);
-        }
-
-        const moduleName = packageConfig.name;
-        const jsiiName = moduleName + '.' + typeName;
-        // Record to typeRefs, so the type's validity is verified.
-        typeRefs.add({ fqn: jsiiName, ctx });
-        return jsiiName;
-    }
-
-    function addDocumentation(target: spec.Documentable, symbol: ts.Symbol) {
-        for (const tag of symbol.getJsDocTags()) {
-            // Don't duplicate @params, they're handled on the params themselves...
-            if (tag.name !== 'param') {
-                target.docs = target.docs || {};
-                target.docs[normalize(tag.name)] = tag.text || '';
-            }
-        }
-
-        const comment = ts.displayPartsToString(symbol.getDocumentationComment(typeChecker));
-        if (comment) {
-            target.docs = target.docs || {};
-            target.docs.comment = comment;
-        }
-
-        function normalize(tagName: string): string {
-            switch (tagName) {
-                // JavaDoc is very particular about it being @return.
-                case 'returns': return 'return';
-                default: return tagName;
+    public async emit(...files: string[]): Promise<EmitResult | never> {
```

Comment:
What's the use case for `files`?

Reply:
It's used when running the tests against the negatives (the standard behaviour otherwise globs all the negatives, which isn't what we want). There was an other way around (basically, copy each file to a temporary work-dir with a `package.json` file to run the tests against), but it was a lot more work.

Adding a comment on `@param` to reflect that.